### ISSUE TITLE
Missing @shareable for PageInfo with federation and the connection spec

### ISSUE THREAD

**Issue by lervag** [ghostdogpr/caliban#2487]
We are using Caliban for a federated subgraph, and we are using the relay connection spec. Thus, we need to have the `@shareable` directive on the `PageInfo` type. We would expect there to be a way to add this directive to the `PageInfo` type from Caliban, but we could not find any simple solutions here.

So, the first part here is to see if we just missed something obvious. If not, then we believe Caliban should automatically add the `@shareable` directive to `PageInfo` for federated graphs.

Second, we have a temporary fix for this with our own hacky schema rewriter. We were curious if there is a relatively simple way to rewrite the resulting schema to add this directive with built-in Caliban features.

**Comment by paulpdaniels** [comment]
You should just be able to add `@GQLShareable` to any type and it will get picked up by derivation.

**Comment by lervag** [comment]
No, I don't think it is that easy, because the `PageInfo` type is defined by Caliban. When implementing this, we relied on the docs here: https://ghostdogpr.github.io/caliban/docs/relay-connections.html.

**Comment by paulpdaniels** [comment]
I think you have two options you could try with the current implementation:

1. [Hand build](https://ghostdogpr.github.io/caliban/docs/schema.html#building-schemas-by-hand) the schema for `PageInfo` instead of generating it
2. Use a `Transformer` similar to the `ExcludeDirectives` transformer (but adding a directive instead based on type name).

**Comment by lervag** [comment]
> I think you have two options you could try with the current implementation: …

Thanks! I'll look into these; the latter seems similar to what we are already doing, but perhaps better and less hacky.

I notice that you are one of the contributors to the federation stuff. So, I'm curious as to what you think of the issue I raise? I mean, should not Caliban add `@shareable` to `PageInfo` automatically if we are building a federated graph?

**Comment by frekw** [comment]
> I notice that you are one of the contributors to the federation stuff. So, I'm curious as to what you think of the issue I raise? I mean, should not Caliban add @shareable to PageInfo automatically if we are building a federated graph?

I think ideally yes, but I'm not sure if ensuring that is feasible.

When you start diving into federation, the spec itself comes with a bunch of assumptions that don't always align perfectly with other assumptions a GraphQL library makes (for example, types that can only be resolved via federated `_entity` queries), so it's _very_ likely that you'll need to make manual tweaks to the schema regardless, so in that light this is probably something you'll have to become comfortable with regardless.

Here are two examples from our (production) code base handling different kinds of fed quirks:

```scala
  implicit val pageInfoSchema: CalibanSchema[Env, caliban.relay.PageInfo]        =
    obj("PageInfo", Some(GQLSharedDescription.Connection.PageInfo), List(Shareable))(implicit ft =>
      List(
        field("hasNextPage", None)(_.hasNextPage),
        field("hasPreviousPage", None)(_.hasPreviousPage),
        field("startCursor", None)(_.startCursor),
        field("endCursor", None)(_.endCursor)
      )
    )
```

and

```scala
  val sdl = graph
    .withAdditionalTypes(
      resolvers
        .map(_.toType)
        .flatMap(Types.collectTypes(_))
        .toList
    )
    .withSchemaDirectives(
      List(Versions.v2_3.toDirective)
    )
    .render
```

So I think my point boils down to this: federation is _extremely_ powerful and a fantastic tool, but also a very advanced use case for GraphQL. So before you decide to embark on that route, make sure you're willing to pay the learning cost of using it!

**Comment by paulpdaniels** [comment]
I’m personally not a huge fan of magic additions. Shareable is itself opt-in and relay is a separate spec from federation, so I don’t know that we’d want to automatically add it. I think where we could go is improving the ability to add directives after creation of the schema. We have a way through TypeVisitors but no way to cleanly connect that to Schema (transformers are one way but they can create logic in two separate places which isn’t always nice).

**Comment by ghostdogpr** [comment]
It can be done pretty easily in your own code by making a custom `Schema` from the automatically-generated one: https://scastie.scala-lang.org/0jwGIocUR0KYF0yAIIXd0Q

```scala
val schema: Schema[Any, PageInfo] = Schema.gen
implicit val customSchema: Schema[Any, PageInfo] = new Schema[Any, PageInfo] {
  def toType(isInput: Boolean, isSubscription: Boolean): __Type =
    schema
      .toType(isInput, isSubscription)
      .copy(directives = Some(List(Directive("shareable"))))
  def resolve(value: PageInfo): Step[Any] = schema.resolve(value)
}
```

**Comment by paulpdaniels** [comment]
I think we could make a convenience method for this potentially `addDirectives`.

**Comment by ghostdogpr** [comment]
Maybe just `mapType` otherwise there are too many potential methods. I actually thought it already existed 😆

**Comment by lervag** [comment]
I notice there are a lot of ways to do this, and I'm learning a lot by seeing different ways. So thanks to you all for sharing!

I think we can close this issue now, as I now agree with @paulpdaniels that it is probably better to handle things like this directly/manually.
